src/rust/scheduler/scheduler.rs

```diff
@@ -84,157 +84,175 @@ pub struct Scheduler {
                 }
             }
             panic!("Could not find a valid task id");
         };
+        new_task_id
+    }
 
-        trace!("insert(): name={:?}, id={:?}, index={:?}", task_name, task_id, index);
-
-        // Add a new page to hold this future's status if the current page is filled.
-        while index >= pages.len() << WAKER_BIT_LENGTH_SHIFT {
-            pages.push(WakerPageRef::default());
+    // If the address space for task ids is close to half full, it will become increasingly difficult to avoid
+    // collisions, so we cap the number of tasks.
+    fn panic_if_too_many_tasks(&self) {
+        let task_ids: Ref<HashMap<u64, usize>> = self.task_ids.borrow();
+        if task_ids.len() > MAX_NUM_TASKS {
+            panic!("Too many concurrent tasks");
         }
-        let (page, subpage_ix): (&WakerPageRef, usize) = {
-            let (pages_ix, subpage_ix) = self.get_page_indexes(index);
-            (&pages[pages_ix], subpage_ix)
-        };
-        page.initialize(subpage_ix);
-        Some(TaskHandle::new(task_id, index, page.clone()))
     }
 
-    /// Computes the page and page offset of a given task based on its total offset.
-    fn get_page_indexes(&self, index: usize) -> (usize, usize) {
-        (index >> WAKER_BIT_LENGTH_SHIFT, index & (WAKER_BIT_LENGTH - 1))
+    fn get_page_index_and_offset(&self, pin_slab_index: usize) -> (usize, usize) {
+        let page_index: usize = pin_slab_index >> WAKER_BIT_LENGTH_SHIFT;
+        let page_offset: usize = pin_slab_index & (WAKER_BIT_LENGTH - 1);
+        (page_index, page_offset)
+    }
+
+    // Add new page(s) to hold this future's status if the current page is filled. This may result in addition of
+    // multiple pages because of the gap between the pin slab index and the current page index.
+    fn add_new_pages_to_accommodate_index_gap(&self, pin_slab_index: usize) {
+        let mut page_ref: RefMut<Vec<WakerPageRef>> = self.page_refs.borrow_mut();
+        while pin_slab_index >= page_ref.len() << WAKER_BIT_LENGTH_SHIFT {
+            page_ref.push(WakerPageRef::default());
+        }
     }
 
     /// Poll all futures which are ready to run again. Tasks in our scheduler are notified when
     /// relevant data or events happen. The relevant event have callback function (the waker) which
     /// they can invoke to notify the scheduler that future should be polled again.
     pub fn poll(&self) {
-        let mut pages: RefMut<Vec<WakerPageRef>> = self.pages.borrow_mut();
+        let mut page_refs: RefMut<Vec<WakerPageRef>> = self.page_refs.borrow_mut();
         let mut tasks: RefMut<PinSlab<Box<dyn Task>>> = self.tasks.borrow_mut();
 
         // Iterate through pages.
-        for page_ix in 0..pages.len() {
-            let (notified, dropped): (u64, u64) = {
-                let page: &mut WakerPageRef = &mut pages[page_ix];
-                (page.take_notified(), page.take_dropped())
+        for page_index in 0..page_refs.len() {
```

Comment:
It's probably fine to treat these as: page in pages or waker_page in waker_pages

Reply:
I called it `page_refs` because we have defined separate WakerPage and WakerPageRef structures in code explicitly. They are not implicit refs as part of the language, so wanted to have that distinction. Also, there is functionality in the WakerPageRef's non-trivial `default()` implementation (which we use to allocate the pages here).

So, I thought it might be better to call them what they are. Do let me know if you think otherwise.